README
This is the "Yet Another Modbus" library, or libyam. It is a Modbus/RTU master
library, using termios to access a serial device. It can issue Modbus/RTU
command packets, and interpret the results.

The packets are formed according to the documentation available at the Modbus
IDA website.

http://www.Modbus-IDA.org

Currently, the following types of packets are allowed:
Read Multiple Registers
Read Multiple Inputs
Write Single Coil
Write Single Register
Read Exception Status
Write Multiple Coils
Write Multiple Registers
Report Slave ID

Note for 64-bit users
---------------------
libtool for 64-bit distros such as Fedora 14 that store 32 and 64 bit libraries
in /usr/lib and /usr/lib64 don't seem to "do the right thing" and choose /usr/lib64
even if --prefix to the configure script is set to /usr. The workaround is to
issue

./configure --prefix=/usr --libdir=/usr/lib64

This puts everything into the right place


ChangeLog
2011-01-01 Jim George <jimgeorge@gmail.com>
Release: 0.2
	* Fixed various mistakes in config files that pointed to old stuff
	* Added functions to return the serial device and to set timeout

2011-01-12	Jim George <jimgeorge@gmail.com>
Release: 0.3
	* Bug fixes to all functions that expect variable length input, prevents
	trashing of the stack
	* Added Read Coils and Read Discrete Inputs functions

2012-02-24	Jim George <jimgeorge@gmail.com>
Release: 0.4
	* Added yam_errorstr function
	* Flipped the order of write_multiple_coils and write_multiple_registers
	arguments

2012-02-27	Jim George <jimgeorge@gmail.com>
Release: 0.4.1
	* Added extra params to yam_modbus_init to set flags for parity,
	handshaking, number of bits per word and number of stop bits


tests/test-rtu.c
/**
\file test-rtu.c
\brief Tests the YAM library functionality
\author Jim George
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <yam/modbus.h>
#include <assert.h>
#include <stdint.h>
#include <time.h>
#include <getopt.h>

enum {
	OPT_DEBUG,
	OPT_TIMEOUT,
	OPT_DEVICE,
	OPT_SLAVEADDR,
	OPT_RUNSTATUS,
	OPT_READCOILS,
	OPT_READDISCRETES,
	OPT_READINPUT,
	OPT_READREGISTER,
	OPT_WRITECOIL,
	OPT_WRITEREGISTER,
	OPT_WRITECOILS,
	OPT_WRITEREGISTERS,
};

char *usage_string =
"Test libyam\n"
"Usage: test-rtu [options] [commands]\n"
"Options:\n"
"--debug: Enable debug mode\n"
"--timeout=val: Set timeout (in milliseconds, default = 1 sec)\n"
"--device=dev[,baudrate[,bits[,par[,stop]]]: Set serial device\n"
"             (default: /dev/ttyUSB0, 57600 bps, 8b, Even parity, 1 stop bit)\n"
"--address=addr: Set slave address\n"
"\n"
"Modbus commands:\n"
"--runstatus: Get the running status\n"
"--readcoils=start[,num]: Read specified coils\n"
"--readdiscretes=start[,num]: Read specified discrete inputs\n"
"--readinput=start[,num]: Read specified inputs\n"
"--readregister=start[,num]: Read specified register\n"
"--writecoil=num,val: Write value to specified coil\n"
"--writeregister=num,val: Write value to specified register\n"
"--writecoils=num,val[,num]: Write value to specified coils\n"
"--writeregisters=num,val[,num]: Write value to specified registers\n"
"\n";

int main(int argc, char *argv[])
{
	struct yam_modbus bs;
	struct yam_modbus *bus = NULL;
	uint16_t regs[100];
	uint8_t coils[100];
	int slave_addr = 0x40;
	int opt_idx, opt_errors = 0, opt;
	int baudrate = 57600;
	int serial_flags = 0;
	char serdev[YAM_MAX_DEVICE_NAME] = "/dev/ttyUSB0";

	bus = &bs;

	static struct option opt_lst[] = {
		{"debug", no_argument, 0, OPT_DEBUG},
		{"timeout", required_argument, 0, OPT_TIMEOUT},
		{"device", required_argument, 0, OPT_DEVICE},
		{"address", required_argument, 0, OPT_SLAVEADDR},
		{"runstatus", no_argument, 0, OPT_RUNSTATUS},
		{"readcoils", required_argument, 0, OPT_READCOILS},
		{"readdiscretes", required_argument, 0, OPT_READDISCRETES},
		{"readinput", required_argument, 0, OPT_READINPUT},
		{"readregister", required_argument, 0, OPT_READREGISTER},
		{"writecoil", required_argument, 0, OPT_WRITECOIL},
		{"writeregister", required_argument, 0, OPT_WRITEREGISTER},
		{"writecoils", required_argument, 0, OPT_WRITECOILS},
		{"writeregisters", required_argument, 0, OPT_WRITEREGISTERS},

		{NULL, 0, 0, 0}
	};

	if (argc == 1) {
		puts(usage_string);
		return -1;
	}

	bzero(bus, sizeof(struct yam_modbus));
	bus->serial = -1;
	yam_modbus_init(serdev, baudrate, YAM_SERIAL_FLAGS_DEFAULT, bus);

	int ret = 0;
	int numregs = 1, start = 40210;
	char *save;
	int value;

	while (-1 != (opt = getopt_long(argc, argv, "dv", opt_lst, &opt_idx))) {
		switch(opt) {
		case OPT_DEBUG:
			if (bus->serial == -1) goto bus_not_initialized;
			yam_debug(bus, 1);
			break;
		case OPT_TIMEOUT:
			if (bus->serial == -1) goto bus_not_initialized;
			yam_set_timeout(bus, strtoul(optarg, NULL, 10));
			break;
		case OPT_DEVICE:
			if (bus->serial == -1) goto bus_not_initialized;
			{
			yam_modbus_close(bus);
			char *delims=", ";
			strncpy(serdev, strtok(optarg, delims), YAM_MAX_DEVICE_NAME);
			char *baudstr = strtok(NULL, delims);
			if (baudstr != NULL) baudrate = strtoul(baudstr, NULL, 10);
			else baudrate = 57600;
			char *bitsstr = strtok(NULL, delims);
			if (bitsstr != NULL) {
				if (!strcasecmp(bitsstr, "8")) {
					serial_flags |= YAM_SERIAL_FLAGS_8BIT;
				}
				else if (!strcasecmp(bitsstr, "7")) {
					serial_flags |= YAM_SERIAL_FLAGS_7BIT;
				}
				else if (!strcasecmp(bitsstr, "6")) {
					serial_flags |= YAM_SERIAL_FLAGS_6BIT;
				}
				else {
					printf("Could not parse number of bits\n");
					return -1;
				}
			}
			else {
				serial_flags |= YAM_SERIAL_FLAGS_8BIT;
			}
			char *parstr = strtok(NULL, delims);
			if (parstr != NULL) {
				if (!strcasecmp(parstr, "N")) {
					serial_flags |= YAM_SERIAL_FLAGS_NO_PARITY;
				}
				else if (!strcasecmp(parstr, "E")) {
					serial_flags |= YAM_SERIAL_FLAGS_EVEN_PARITY;
				}
				else if (!strcasecmp(parstr, "O")) {
					serial_flags |= YAM_SERIAL_FLAGS_ODD_PARITY;
				}
				else {
					printf("Could not parse parity type\n");
					return -1;
				}
			}
			else {
				serial_flags |= YAM_SERIAL_FLAGS_NO_PARITY;
			}
			char *stopstr = strtok(NULL, delims);
			if (stopstr != NULL) {
				if (!strcasecmp(stopstr, "1")) {
					serial_flags |= YAM_SERIAL_FLAGS_ONE_STOP;
				}
				else if (!strcasecmp(stopstr, "2")) {
					serial_flags |= YAM_SERIAL_FLAGS_TWO_STOP;
				}
				else {
					printf("Could not parse number of stop bits\n");
					return -1;
				}
			}
			else {
				serial_flags |= YAM_SERIAL_FLAGS_ONE_STOP;
			}
			if (0 > yam_modbus_init(serdev, baudrate, serial_flags , bus)) {
				printf("Error initializing bus with device %s at %d bps\n",
					serdev, baudrate);
				return -1;
			}
			}
			break;
		case OPT_SLAVEADDR:
			slave_addr = strtoul(optarg, NULL, 16);
			break;
		case OPT_RUNSTATUS:
			if (bus->serial == -1) goto bus_not_initialized;
			{
			uint8_t id, run_status, exception_status;
			char additional_data[256];
			int adl;
			ret = yam_report_slave_id(bus, slave_addr, &id, &run_status, additional_data, &adl);
			if (0 > ret) {
				yam_perror(bus, "Error reading Slave ID");
			}
			else {
				printf("ID: %02X, Run status: %02X, %d additional bytes\n", id, run_status, adl);
			}
			ret = yam_read_exception_status(bus, slave_addr, &exception_status);
			if (0 > ret) {
				yam_perror(bus, "Error reading exception status");
			}
			else {
				printf("Exception status: %02X (%d)\n", exception_status, exception_status);
			}
			}
			break;
		case OPT_READCOILS:
			if (bus->serial == -1) goto bus_not_initialized;
			start = strtoul(optarg, &save, 10);
			if (0 != *save) numregs = strtoul(save + 1, NULL, 10);
			else numregs = 1;
			ret = yam_read_coils(bus, slave_addr, start, numregs, coils);
			if (0 > ret) {
				yam_perror(bus, "Error reading coils");
			}
			else {
				int ctr;
				for (ctr = 0; ctr < numregs; ctr++) {
					printf("Coil %d = %s\n", ctr + start, coils[ctr] ? "ON" : "OFF");
				}
			}
			break;
		case OPT_READDISCRETES:
			if (bus->serial == -1) goto bus_not_initialized;
			start = strtoul(optarg, &save, 10);
			if (0 != *save) numregs = strtoul(save + 1, NULL, 10);
			else numregs = 1;
			ret = yam_read_discretes(bus, slave_addr, start, numregs, coils);
			if (0 > ret) {
				yam_perror(bus, "Error reading discrete inputs");
			}
			else {
				int ctr;
				for (ctr = 0; ctr < numregs; ctr++) {
					printf("Discrete input %d = %s\n", ctr + start, coils[ctr] ? "ON" : "OFF");
				}
			}
			break;
		case OPT_READINPUT:
			if (bus->serial == -1) goto bus_not_initialized;
			start = strtoul(optarg, &save, 10);
			if (0 != *save) numregs = strtoul(save + 1, NULL, 10);
			else numregs = 1;
			ret = yam_read_inputs(bus, slave_addr, start, numregs, regs);
			if (0 > ret) {
				yam_perror(bus, "Error reading registers");
			}
			else {
				int ctr;
				for (ctr = 0; ctr < numregs; ctr++) {
					printf("Input %d = %d (0x%04X)\n", ctr + start, regs[ctr], regs[ctr]);
				}
			}
			break;
		case OPT_READREGISTER:
			if (bus->serial == -1) goto bus_not_initialized;
			start = strtoul(optarg, &save, 10);
			if (0 != *save) numregs = strtoul(save + 1, NULL, 10);
			else numregs = 1;
			ret = yam_read_registers(bus, slave_addr, start, numregs, regs);
			if (0 > ret) {
				yam_perror(bus, "Error reading registers");
			}
			else {
				int ctr;
				for (ctr = 0; ctr < numregs; ctr++) {
					printf("Input %d = %d (0x%04X)\n", ctr + start, regs[ctr], regs[ctr]);
				}
			}
			break;
		case OPT_WRITECOIL:
			if (bus->serial == -1) goto bus_not_initialized;
			start = strtoul(optarg, &save, 10);
			if (0 != *save) value = strtoul(save + 1, NULL, 10);
			else value = 0;
			ret = yam_write_single_coil(bus, slave_addr, start, value);
			if (0 > ret) {
				yam_perror(bus, "Error writing coil");
			}
			break;
		case OPT_WRITEREGISTER:
			if (bus->serial == -1) goto bus_not_initialized;
			start = strtoul(optarg, &save, 10);
			if (0 != *save) value = strtoul(save + 1, NULL, 16);
			else value = 0;
			ret = yam_write_single_register(bus, slave_addr, start, value);
			if (0 > ret) {
				yam_perror(bus, "Error writing register");
			}
			break;
		case OPT_WRITECOILS:
			if (bus->serial == -1) goto bus_not_initialized;
			start = strtoul(optarg, &save, 10);
			if (0 != *save) value = strtoul(save + 1, &save, 10);
			else value = 0;
			if (0 != *save) numregs = strtoul(save + 1, NULL, 10);
			else numregs = 1;
			for (ret = 0; ret < numregs; ret++) {
				coils[ret] = value;
			}
			ret = yam_write_multiple_coils(bus, slave_addr, start, numregs, coils);
			if (0 > ret) {
				yam_perror(bus, "Error writing coils");
			}
			break;
		case OPT_WRITEREGISTERS:
			if (bus->serial == -1) goto bus_not_initialized;
			start = strtoul(optarg, &save, 10);
			if (0 != *save) value = strtoul(save + 1, &save, 10);
			else value = 0;
			if (0 != *save) numregs = strtoul(save + 1, NULL, 16);
			else numregs = 1;
			for (ret = 0; ret < numregs; ret++) {
				regs[ret] = value;
			}
			ret = yam_write_multiple_registers(bus, slave_addr, start, numregs, regs);
			if (0 > ret) {
				yam_perror(bus, "Error writing register");
			}
			break;
		default:
			opt_errors++;
			break;
		}
	};
	if (opt_errors) {
		puts(usage_string);
	}
	yam_modbus_close(bus);

	return 0;

bus_not_initialized:
	printf("Default device (%s) not available\n", serdev);
	puts ("Serial bus not initialized");
	return -1;
}



yam/serial.c
/**
\file serial.c
\brief Module for YAM serial port handling
\author Jim George
*/

#include <unistd.h>
#include <stdio.h>
#include <strings.h>
#include <assert.h>
#include <stdlib.h>
#include <fcntl.h>
#include <termios.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <linux/serial.h>
#include <errno.h>
#include "serial.h"
#include "modbus.h"

#define SERIAL_PORT_SPD_TBL_MAX 18
static struct {
	int baud;
	speed_t ident;
} serial_port_speed_table[SERIAL_PORT_SPD_TBL_MAX] = {
{50, B50},
{75, B75},
{110, B110},
{134, B134},
{150, B150},
{200, B200},
{300, B300},
{600, B600},
{1200, B1200},
{1800, B1800},
{2400, B2400},
{4800, B4800},
{9600, B9600},
{19200, B19200},
{38400, B38400},
{57600, B57600},
{115200, B115200},
{230400, B230400}
};

/**
\brief Get interface speed macro from integer speed
\param speed Input speed, in baud
\return One of the interface speed constants from termios library,
or -1 on error

Converts an integer baud rate into one of the termios speed
macros. If the baud rate specified is not found, an error is
returned.
*/
static speed_t serial_port_get_speed(const unsigned int speed)
{
	int ctr;

	for (ctr = 0; ctr < SERIAL_PORT_SPD_TBL_MAX; ctr++) {
		if (speed == serial_port_speed_table[ctr].baud) {
			return serial_port_speed_table[ctr].ident;
		}
	}
	return -1;
}

/**
\brief Initialize serial port
\param *device_name Unix device name for the serial port to open
\param speed Speed at which to open port
\param *port Pointer to the file descriptor for the serial port
\return 0 on success, -1 on error

The port is opened with 8N1 settings (8-bit, no parity, 1 stop bit)
*/
int serial_port_init(const char *device_name,
	unsigned int speed, unsigned int flags,
	int *port)
{
	struct termios term_st;

	assert(device_name != NULL);
	assert(port != NULL);

	*port = open(device_name, O_RDWR | O_NOCTTY);
	if (*port < 0) {
		return -1;
	}

	bzero(&term_st, sizeof(struct termios));

	if (tcgetattr(*port, &term_st)) {
		perror("tcgetattr");
		return -1;
	}

	int status;
	ioctl(*port, TIOCMGET, &status);
	status &= ~TIOCM_DTR;
	ioctl(*port, TIOCMSET, &status);

	/* Set interface speed */
	{
		int spd_macro = serial_port_get_speed(speed);
		if (spd_macro == -1) return -1;
		cfsetispeed(&term_st, spd_macro);
		cfsetospeed(&term_st, spd_macro);
	}
	/* Enable raw mode output */
	cfmakeraw(&term_st);
	term_st.c_oflag &= ~OPOST;

	/* Don't own the terminal (^C won't kill us), Enable rcv, Drop DTR on close */
	term_st.c_cflag |= (CREAD | HUPCL | CLOCAL);

	/* Set 8-bit port size */
	term_st.c_cflag &= ~CSIZE;
	switch (flags & YAM_SERIAL_FLAGS_BITS_MSK) {
	case YAM_SERIAL_FLAGS_8BIT:
		term_st.c_cflag |= CS8;
		break;
	case YAM_SERIAL_FLAGS_7BIT:
		term_st.c_cflag |= CS7;
		break;
	case YAM_SERIAL_FLAGS_6BIT:
		term_st.c_cflag |= CS6;
		break;
	}

	/* No parity, 1 stop bit */
	term_st.c_cflag &= ~(PARENB | PARODD);
	switch (flags & YAM_SERIAL_FLAGS_PARITY_MSK) {
	case YAM_SERIAL_FLAGS_NO_PARITY:
	default:
		break;
	case YAM_SERIAL_FLAGS_ODD_PARITY:
		term_st.c_cflag |= (PARENB | PARODD);
		break;
	case YAM_SERIAL_FLAGS_EVEN_PARITY:
		term_st.c_cflag |= PARENB;
		break;
	}
	if (flags & YAM_SERIAL_FLAGS_TWO_STOP) {
		term_st.c_cflag |= CSTOPB;
	}
	else {
		term_st.c_cflag &= ~CSTOPB;
	}

	/* Turn off flow control, RS-485 does not use it */
	term_st.c_cflag &= ~CRTSCTS;
	term_st.c_iflag &= ~(IXON | IXOFF | IXANY);
	switch (flags & YAM_SERIAL_FLAGS_HANDSHAKE_MSK) {
	case YAM_SERIAL_FLAGS_NO_HANDSHAKE:
	default:
		break;
	case YAM_SERIAL_FLAGS_HW_HANDSHAKE:
		term_st.c_cflag |= CRTSCTS;
		break;
	case YAM_SERIAL_FLAGS_SW_HANDSHAKE:
		term_st.c_iflag |= (IXON | IXOFF);
		break;
	}

	/* Set timeout to 1 second */
	term_st.c_cc[VMIN] = 0;
	term_st.c_cc[VTIME] = 0;

	if (tcflush(*port, TCIOFLUSH)) {
		perror("tcflush");
		return -1;
	}
	if (tcsetattr(*port, TCSANOW, &term_st)) {
		perror("tcgetattr");
		return -1;
	}

	ioctl(*port, TIOCMGET, &status);
	status |= TIOCM_DTR;
	ioctl(*port, TIOCMSET, &status);

	return 0;
}

/**
\brief Flush serial port
\param fd Device number of serial port to flush

Flush the buffers of the serial port
*/
void serial_port_flush(int fd)
{
	tcflush(fd, TCIOFLUSH);
}


yam/modbus.h
/**
\file modbus.h
\brief Include file for YAM - Yet Another Modbus/RTU library
\author Jim George
*/

#ifndef _YAM_MODBUS_H_
#define _YAM_MODBUS_H_

#include <stdint.h>

#define YAM_MAX_DEVICE_NAME 64

/**
\brief The YAM object

This structure represents the YAM object. One may be created using the
yam_modbus_init function. All calls to the YAM library require a YAM object,
to know which modbus serial device to use. This permits multiple serial
devices to be open simultaneously.
*/
struct yam_modbus {
	int serial; /**< Serial port file descriptor */
	int baudrate; /**< Baud rate */
	int debug; /**< Nonzero to enable debug stuff to stdout */
	int timeout_ms; /**< Timeout, in milliseconds, when reading */
	int last_errorcode; /**< Last error code seen by this bus */
	char device_name[YAM_MAX_DEVICE_NAME]; /**< Name of the serial device */
	char slaveidhack; /**< Set nonzero to subtract 1 from slave ID additional bytes
	                       field, to match nonstandard behavior of libmodbus */
};

/* Serial flags */
#define YAM_SERIAL_FLAGS_PARITY_MSK (3 << 0)
#define YAM_SERIAL_FLAGS_NO_PARITY (0 << 0)
#define YAM_SERIAL_FLAGS_EVEN_PARITY (1 << 0)
#define YAM_SERIAL_FLAGS_ODD_PARITY (2 << 0)
#define YAM_SERIAL_FLAGS_HANDSHAKE_MSK (3 << 2)
#define YAM_SERIAL_FLAGS_NO_HANDSHAKE 0
#define YAM_SERIAL_FLAGS_HW_HANDSHAKE (1 << 2)
#define YAM_SERIAL_FLAGS_SW_HANDSHAKE (2 << 2)
#define YAM_SERIAL_FLAGS_ONE_STOP 0
#define YAM_SERIAL_FLAGS_TWO_STOP (1 << 4)
#define YAM_SERIAL_FLAGS_BITS_MSK (3 << 5)
#define YAM_SERIAL_FLAGS_8BIT (0 << 5)
#define YAM_SERIAL_FLAGS_6BIT (1 << 5)
#define YAM_SERIAL_FLAGS_7BIT (2 << 5)

#define YAM_SERIAL_FLAGS_8N1 (YAM_SERIAL_FLAGS_8BIT | \
		YAM_SERIAL_FLAGS_NO_HANDSHAKE | \
		YAM_SERIAL_FLAGS_NO_PARITY | \
		YAM_SERIAL_FLAGS_ONE_STOP)

#define YAM_SERIAL_FLAGS_DEFAULT (YAM_SERIAL_FLAGS_8BIT | \
		YAM_SERIAL_FLAGS_NO_HANDSHAKE | \
		YAM_SERIAL_FLAGS_EVEN_PARITY | \
		YAM_SERIAL_FLAGS_ONE_STOP)

/* MODBUS Function codes */
#define YAM_READ_COILS 0x01
#define YAM_READ_DISCRETES 0x02
#define YAM_READ_REGISTERS 0x03
#define YAM_READ_INPUTS 0x04
#define YAM_WRITE_SINGLECOIL 0x05
#define YAM_WRITE_SINGLEREGISTER 0x06
#define YAM_READ_EXCEPTIONSTATUS 0x07
/*
#define YAM_READ_DIAGNOSTICS 0x08
#define YAM_READ_COMMEVTCOUNTER 0x0B
#define YAM_READ_COMMEVTLOG 0x0C
*/
#define YAM_WRITE_COILS 0x0F
#define YAM_WRITE_REGISTERS 0x10
#define YAM_REPORTSLAVEID 0x11

/* MODBUS Exception codes */
/** Return code - slave does not respond to specified function code */
#define YAM_ILLEGAL_FUNCTION -1
/** Return code - slave reports invalid address */
#define YAM_ILLEGAL_DATA_ADDR -2
/** Return code - slave reports that value written to register is invalid */
#define YAM_ILLEGAL_DATA_VALUE -3
/** Return code - slave reports a general failure */
#define YAM_SLAVE_FAILURE -4
/** Return code - slave acknowledges the request but did not process it */
#define YAM_ACKNOWLEDGE -5
/** Return code - slave is busy and cannot process the request */
#define YAM_SLAVE_BUSY -6
/** Return code - command sent to slave had a parity or CRC error */
#define YAM_PARITY_ERROR -8

/* Exeception codes returned by YAM */
/** Return code - everything's OK */
#define YAM_OK 0
/** Return code - response from slave had a bad CRC */
#define YAM_CRC_ERROR -256
/** Return code - slave timed out */
#define YAM_TIMEOUT -257
/** Return code - invalid number of bytes returned by slave */
#define YAM_INVALIDBYTECOUNT -258
/** Return code - serial port init failed */
#define YAM_SERIAL_INIT_FAILED -259
/** Return code - too many registers/coils (exceeds ADU size) */
#define YAM_TOO_MANY_REGISTERS -260

/** Maximum ADU length, in bytes */
#define YAM_MODBUS_MAX_ADU_LEN 256
/** Maximum PDU length, in bytes */
#define YAM_MODBUS_MAX_PDU_LEN 253
/** Maximum number of registers per request */
#define YAM_REGS_PER_REQUEST 123
/** Maximum number of coils per request */
#define YAM_COILS_PER_REQUEST 1968
/** Default timeout of a request, in milliseconds */
#define YAM_DEFAULT_TIMEOUT 1000

int yam_modbus_init(const char *device_name,
             unsigned int speed, unsigned int flags,
             struct yam_modbus *bus);
void yam_modbus_close(struct yam_modbus *bus);
void yam_debug(struct yam_modbus *bus, int debug_status);
void yam_set_timeout(struct yam_modbus *bus, int timeout_ms);
int yam_get_serial_device(struct yam_modbus *bus);

int yam_read_coils(struct yam_modbus *bus, uint8_t addr,
                       uint16_t start_addr, uint16_t num_coils,
                       uint8_t *coils);
int yam_read_discretes(struct yam_modbus *bus, uint8_t addr,
                       uint16_t start_addr, uint16_t num_discretes,
                       uint8_t *discretes);
int yam_read_registers(struct yam_modbus *bus, uint8_t addr,
                       uint16_t start_addr, uint16_t num_regs,
                       uint16_t *regs);
int yam_read_inputs(struct yam_modbus *bus, uint8_t addr,
                       uint16_t start_addr, uint16_t num_regs,
                       uint16_t *regs);
int yam_write_single_coil(struct yam_modbus *bus, uint8_t addr,
                       uint16_t coil_addr, uint8_t coil_state);
int yam_write_single_register(struct yam_modbus *bus, uint8_t addr,
                       uint16_t register_addr, uint16_t register_value);
int yam_read_exception_status(struct yam_modbus *bus, uint8_t addr,
                              uint8_t *exception_status);
int yam_write_multiple_coils(struct yam_modbus *bus, uint8_t addr,
                             uint16_t start_addr, uint16_t num_coils,
                             uint8_t *coils);
int yam_write_multiple_registers(struct yam_modbus *bus, uint8_t addr,
                                 uint16_t start_addr, uint16_t num_regs,
                                 uint16_t *regs);
int yam_report_slave_id(struct yam_modbus *bus, uint8_t addr, uint8_t *id,
                        uint8_t *run_status, char *additional_data, int *buflen);

void yam_perror(struct yam_modbus *bus, char *s);
char *yam_strerror(int errnum);
char *yam_errorstr(struct yam_modbus *bus);

#endif /* _YAM_MODBUS_H_ */
